src/app/qgisapp.cpp

```diff
@@ -1941,6 +1954,131 @@ QgsMessageBar *QgisApp::visibleMessageBar()
   }
 }
 
+QList<QgsVectorLayerRef> QgisApp::findBrokenWidgetDependencies( QgsVectorLayer *vl )
+{
+  QList<QgsVectorLayerRef> brokenDependencies;
+  // Check for missing layer widget dependencies
+  for ( int i = 0; i < vl->fields().count(); i++ )
+  {
+    std::unique_ptr<QgsEditorWidgetWrapper> ww;
+    ww.reset( QgsGui::editorWidgetRegistry()->create( vl, i, nullptr, nullptr ) );
```

Comment:
@elpaso I get a crash (failing assert) since this change here when loading a project (qgep).

![image](https://user-images.githubusercontent.com/588407/70787927-10a58300-1d90-11ea-872e-3808cfc20cbe.png)

What I wonder now is if the approach to *create* a disposable widget wrapper here is the right way to go. Or if this kind of handling would be better suited in the factory directly (maybe even in the fieldformatter in core since it's not really GUI related and could be used by qfield or theoretically even server too).

Reply:
Do you mean moving `QgsEditorWidgetWrapper::layerDependencies()` logic into the field formatter?

I'm not sure that it's the right place, the configuration about widgets "dependencies" is stored in the widget configuration and it belongs to the widget.

What exactly are you suggesting?

Isn't actually better to check if the assert can be removed putting some safe guards wherever in the widget logic there is a danger of misbehavior due to vectorLayerTools being nullptr?

Or even easier: create the`QgsAttributeEditorContext` and pass it to the factory?

Reply:
My preferred approach would indeed be to move it to the field formatter.

The field formatter is also based on the same configuration (see e.g. https://github.com/qgis/QGIS/blob/master/src/core/fieldformatter/qgsvaluerelationfieldformatter.cpp#L119 ) and lives in core so the logic can be used everywhere. They are created once per application lifetime, so there is no need to create disposable objects.

But, since I'm not completely sure that there are no hidden dependencies involved, the second proposal less risky but also less generic would be to move it to the widget factories. They are also only created once per application lifetime.

Your proposal will be the easiest in terms of work (because it follows the structure set up in this PR), but is also mostly putting the dust under the carpet, to resolve layer dependencies there is no widget (gui element) required, so I don't think creating a throw away widget wrapper for this purpose is the best way to implement it.

Reply:
Ok, I'll look into the field formatter approach.

Reply:
Thanks.
Let me know if you run into questions!